py/picca/pk1d/postproc_pk1d.py

```diff
@@ -205,229 +201,599 @@ def compute_mean_pk1d(p1d_table,
-                delta_z = delta_z[0]
 
-                select = (p1d_table['k'] < kbin_edges[ikbin + 1]) & (
-                            p1d_table['k'] > kbin_edges[ikbin]
-                        )
-                if izbin in (0, nbins_z-1):
-                    # First and last bin: in order to avoid edge effects,
-                    #    use only chunks within the bin
-                    select = select & (
-                        p1d_table['forest_z'] > zbin_edges[izbin]) & (
-                        p1d_table['forest_z'] < zbin_edges[izbin+1])
-                else:
-                    select = select & (
-                        p1d_table['forest_z'] < zbin_centers[izbin+1]) & (
-                        p1d_table['forest_z'] > zbin_centers[izbin-1])
+    Arguments
+    ---------
+    p1d_table: numpy ndarray,
+    Table containing the data to be averaged.
 
-                redshift_weights = 1.0 - np.abs(p1d_table['forest_z'][select]-
-                                                zbin_centers[izbin]) / delta_z
+    p1d_table_cols: List of str,
+    Column names in the input table to be averaged.
 
+    mean_p1d_table: astropy Table,
+    Output table to be filled with the mean values.
+
+    weight_method: str,
+    Method to weight the data.
+
+    apply_z_weights: bool,
+    If True, apply redshift weights.
+
+    snrfit_table: numpy ndarray,
+    Table containing the fit to the SNR.
+
+    output_snrfit: bool,
+    If not None, write the fit to the SNR to a file.
+
+    nomedians: bool,
+    If True, do not use median values in the fit to the SNR.
+
+    nbins_z: int,
+    Number of redshift bins.
+
+    zbin_centers: numpy ndarray,
+    Centers of the redshift bins.
+
+    zbin_edges: numpy ndarray,
+    Edges of the redshift bins.
+
+    n_chunks: numpy ndarray,
+    Number of chunks in each redshift bin.
+
+    nbins_k: int,
+    Number of k bins.
+
+    kbin_edges: numpy ndarray,
+    Edges of the k bins.
+
+    izbin: int,
+    Index of the current redshift bin.
+
+    Return
+    ------
+    None
+    """
+
+    if n_chunks[izbin] == 0:  # Fill rows with NaNs
+        i_min = izbin * nbins_k
+        i_max = (izbin + 1) * nbins_k
+        mean_p1d_table["zbin"][i_min:i_max] = zbin_centers[izbin]
+        mean_p1d_table["index_zbin"][i_min:i_max] = izbin
+        return
+
+    for ikbin, kbin in enumerate(kbin_edges[:-1]):  # Main loop 2) k bins
+
+        if apply_z_weights:  # special chunk selection in that case
+            delta_z = zbin_centers[1:] - zbin_centers[:-1]
+            if not np.allclose(delta_z, delta_z[0], atol=1.0e-3):
+                raise ValueError(
+                    "z bins should have equal widths with apply_z_weights."
+                )
+            delta_z = delta_z[0]
+
+            select = (p1d_table["k"] < kbin_edges[ikbin + 1]) & (
+                p1d_table["k"] > kbin_edges[ikbin]
+            )
+            if izbin in (0, nbins_z - 1):
```

Comment:
Not sure I understand your question. In general, select is on edges.
At this specific point of the code, there is an exception for z bin selection, when apply_z_weights is True:
in this case, since there is triangle-shape z-weighting ("a la QMLE") you need to extend the selection to nearest bins; on the edges, this leads to a bias (ie for z bin "0" you'll have data from z bin "1" but no data from z bin "-1"...) => there is an exception.
Well, maybe that wasn't your question.

Reply:
ah, I see. So the selection in the weighted case  is such that for bin i you'd select data from the center of bin i-1 to the center of bin i+1, in total spanning 2 full bins around the center of bin i, with a weighting scheme downweighting far away points.
In that case this should be completely fine, just didn't properly read what is going on I guess...
How is the weighting done for the corner cases? Do we get a bias there from the asymmetrically distributed data wrt the bin center...

Reply:
ah, got it. So the boundary bins only take the data from within that bin, but elsewise apply the same weighting scheme as other bins which additionally take half a bin outside in each direction into account. Which removes the asymmetry and therefore also the biasing.
Thanks for those explanations!